apps/framework-docs/src/pages/moose/building/consumption-apis.mdx

```diff
@@ -511,7 +520,92 @@ The `sql` template literal in Moose provides type-safe query construction with p
 </TypeScript>
 
 <Python>
-The MooseClient enables you to execute queries and other database operations. Specifically by using `f-string` interpolation to build your queries.
+
+There are two recommended patterns for building your queries:
+
+<Tabs items={["Using Parameter Binding", "Using F-strings"]}>
+<Tabs.Tab title="Using Parameter Binding">
+
+**Pattern 1: ClickHouse Native Parameter Binding**
+
+This pattern follows the official ClickHouse Python client approach using parameter binding:
+
+```python filename="ExampleApi.py" copy
+def run(client: MooseClient, params: QueryParams):
+    # Use named parameters with {:type} syntax for safe parameter binding
+    query = """
+        SELECT *
+        FROM users
+        WHERE age >= {min_age:Int32}
+        AND status = {status:String}
+        LIMIT {limit:UInt32}
+    """
+    
+    # Parameters passed as a dictionary
+    result = client.query.execute(
+        query,
+        parameters={
+            "min_age": params.min_age,
+            "status": params.status,
+            "limit": params.limit
+        }
+    )
+    return result
+```
+
+This approach:
+- Uses ClickHouse's native parameter binding with `{:type}` syntax
+- Explicitly specifies ClickHouse types for parameters
+- Provides protection against SQL injection by default
+- Handles type conversion automatically
+
+</Tabs.Tab>
+
+<Tabs.Tab title="Using F-strings">
+
+**Pattern 2: F-String with Pydantic Validation**
+
+Alternatively, you can use f-strings with careful Pydantic validation:
+
+```python filename="ExampleApi.py" copy
+from pydantic import BaseModel, Field, constr
+from typing import Literal
+
+class QueryParams(BaseModel):
+    # Strict validation ensures safety when using f-strings
+    min_age: int = Field(ge=0, le=150)  # Constrain age range
+    status: str = Field(pattern=r"^(active|inactive)$") ## Only allow specific values
+    limit: int = Field(default=10, ge=1, le=1000)  # Constrain limit range
+    search_text: str = Field(pattern=r'^[a-zA-Z0-9\s]*$')  # Only alphanumeric + spaces
+
+def run(client: MooseClient, params: QueryParams):
+    # Since Pydantic has validated the inputs, f-string usage is safe
+    query = f"""
+        SELECT *
+        FROM users
+        WHERE age >= {params.min_age}
+        AND status = '{params.status}'
+        AND name ILIKE '%{params.search_text}%'
+        LIMIT {params.limit}
+    """
+    
+    return client.query.execute(query)
+```
+</Tabs.Tab>
+</Tabs>
+
+<Callout type="warning">
+When using Pattern 2 (f-strings), you MUST be extremely careful with your Pydantic model definitions:
```

Comment:
i think an objection can be, why bother mentioning f-strings at all - because it makes the demos a little nicer to at?